File: project1/assign1/src/assignement_1.cpp
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <time.h>
#include <cstdlib>

using namespace std;

#define SYSTEMTIME clock_t

 
void OnMult(int m_ar, int m_br) 
{
    SYSTEMTIME Time1, Time2;
    
    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i = 0; i < m_ar; i++)
        for(j = 0; j < m_ar; j++)
            pha[i*m_ar + j] = 1.0;

    for(i = 0; i < m_br; i++)
        for(j = 0; j < m_br; j++)
            phb[i*m_br + j] = (double)(i + 1);

    Time1 = clock();

    for(i = 0; i < m_ar; i++)
    {
        for(j = 0; j < m_br; j++)
        {
            temp = 0;
            for(k = 0; k < m_ar; k++)
            {    
                temp += pha[i*m_ar + k] * phb[k*m_br + j];
            }
            phc[i*m_ar + j] = temp;
        }
    }

    Time2 = clock();
    snprintf(st, sizeof(st), "Time: %3.3f seconds\n",
         (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    cout << "Result matrix: " << endl;
    for(i = 0; i < 1; i++)
    {
        for(j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}
 

// Line-by-line matrix multiplication
double OnMultLine(int m_ar, int m_br)
{
    SYSTEMTIME Time1, Time2;
    
    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for(i = 0; i < m_ar; i++){
        for(j = 0; j < m_ar; j++){
            pha[i*m_ar + j] = 1.0;
            phc[i*m_ar + j] = 0.0;
        }
    }

    for(i = 0; i < m_br; i++){
        for(j = 0; j < m_br; j++){
            phb[i*m_br + j] = (double)(i + 1);
        }

    }

    
    Time1 = clock();

    for(i = 0; i < m_ar; i++)
    {
        for(j = 0; j < m_ar; j++)
        {
            for (k=0; k < m_ar; k++){
                phc[i*m_ar + k]+= pha[i*m_ar + j]*phb[j*m_ar + k];
            }
        }
    }

    Time2 = clock();
    double elapsed = (double)(Time2 - Time1) / CLOCKS_PER_SEC;
    snprintf(st, sizeof(st), "Time: %3.3f seconds\n", elapsed);
    cout << st;

    cout << "Result matrix: " << endl;
    for(i = 0; i < 1; i++)
    {
        for(j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    return elapsed;
   
}


// Block matrix multiplication
void OnMultBlock(int m_ar, int m_br, int bkSize)
{
    // assuming m_ar == m_br
    int N = m_ar;

    clock_t Time1, Time2;
    char st[100];

    double *pha, *phb, *phc;

    pha = (double*)malloc(N * N * sizeof(double));
    phb = (double*)malloc(N * N * sizeof(double));
    phc = (double*)malloc(N * N * sizeof(double));

    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            pha[i * N + j] = 1.0;

    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            phb[i * N + j] = (double)(i + 1);

    for (int i = 0; i < N * N; i++)
        phc[i] = 0.0;

    Time1 = clock();

    for (int ii = 0; ii < N; ii += bkSize) {
        int iMax = std::min(ii + bkSize, N);
        for (int kk = 0; kk < N; kk += bkSize) {
            int kMax = std::min(kk + bkSize, N);
            for (int jj = 0; jj < N; jj += bkSize) {
                int jMax = std::min(jj + bkSize, N);
                for (int i = ii; i < iMax; i++) {
                    for (int k = kk; k < kMax; k++) {
                        double a = pha[i * N + k];
                        for (int j = jj; j < jMax; j++) {
                            phc[i * N + j] += a * phb[k * N + j];
                        }
                    }
                }
            }
        }
    }

    Time2 = clock();

    snprintf(st, sizeof(st), "Time: %3.3f seconds\n",
             (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    cout << "Result matrix: " << endl;
    for (int j = 0; j < std::min(10, N); j++)
        cout << phc[j] << " ";
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}


int main(int argc, char *argv[]) {
    if (argc < 3) {
        cerr << "Usage: " << argv[0] << " <matrix_size> " << endl;
        return 1;
    }
    int size   = atoi(argv[1]);
    int bkSize = atoi(argv[2]);

    cout << "Running OnMultBlock with size " << size << "x" << size
        << endl;
    OnMultLine(size, size);
    return 0;
}

File: project1/assign1/src/assignement_1.py
import time

def on_mult(m_ar, b_ar):
    pha = [0.0] * (m_ar * m_ar)
    phb = [0.0] * (b_ar * b_ar)
    phc = [0.0] * (m_ar * m_ar)

    # Fill A with 1.0
    for i in range(m_ar):
        for j in range(b_ar):
            pha[i * m_ar + j] = 1.0

    # Fill B with (i + 1)
    for i in range(b_ar):
        for j in range(b_ar):
            phb[i * m_ar + j] = float(i + 1) 

    start = time.perf_counter()

    for i in range(m_ar):
        for j in range(m_ar):
            temp = 0.0
            for k in range(m_ar):
                temp += pha[i * m_ar + k] * phb[k * b_ar + j]
            phc[i * m_ar + j] = temp

    end = time.perf_counter()

    print("Time:", end - start, "seconds")
    print("Result matrix:")

    # Print first row, first min(10, b_ar) elements (like your JS)
    for j in range(min(10, b_ar)):
        print(phc[j], end=" ")
    print()

def onMultLine(m_ar, m_br):
    pha = [0.0] * (m_ar * m_ar)
    phb = [0.0] * (m_br * m_br)
    phc = [0.0] * (m_ar * m_ar)

    # Fill A with 1.0
    for i in range(m_ar):
        for j in range(m_br):
            pha[i * m_ar + j] = 1.0

    # Fill B with (i + 1)
    for i in range(m_br):
        for j in range(m_br):
            phb[i * m_ar + j] = float(i + 1) 
    
    start = time.perf_counter()
    
    for i in range (m_br):
        for k in range (m_ar):
            temp = pha[i*m_ar + k]
            for j in range (m_br):
                phc[i*m_ar + j] += temp * phb[k*m_ar+j]

    end = time.perf_counter()

    print("Time:", end - start, "seconds")
    print("Result matrix:")

    for j in range(min(10, m_br)):
        print(phc[j], end=" ")
    print()

def onMultBlock(m_ar, m_br, bk_size):
    pha = [0.0] * (m_ar * m_ar)
    phb = [0.0] * (m_br * m_br)
    phc = [0.0] * (m_ar * m_ar)

    # Fill A with 1.0
    for i in range(m_ar):
        for j in range(m_br):
            pha[i * m_ar + j] = 1.0

    # Fill B with (i + 1)
    for i in range(m_br):
        for j in range(m_br):
            phb[i * m_ar + j] = float(i + 1) 

    start = time.perf_counter()


    for ii in range(0, m_ar, bk_size):
        iMax = min(ii + bk_size, m_ar)
        for kk in range(0, m_ar, bk_size):
            kMax = min(kk + bk_size, m_ar)
            for jj in range(0, m_ar, bk_size):
                jMax = min(jj + bk_size, m_ar)
                for i in range(ii, iMax):
                    iN = i * m_ar
                    for k in range(kk, kMax):
                        a = pha[iN + k]
                        kN = k * m_ar
                        for j in range(jj, jMax):
                            phc[iN + j] += a * phb[kN + j]

    end = time.perf_counter()

    print("Time:", end - start, "seconds")
    print("Result matrix:")

    for j in range(min(10, m_br)):
        print(phc[j], end=" ")
    print()

onMultLine(3072, 3072)7